src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -534,30 +535,48 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
 
       val lazyName  = lazyVal.name.toTermName
       val pos       = lazyVal.pos.focus
-      val flags     = (lazyVal.flags & FieldFlags | ARTIFACT | MUTABLE) & ~(IMPLICIT | STABLE) // TODO: why include MUTABLE???
-      val holderSym = owner.newValue(lazyName append nme.LAZY_LOCAL_SUFFIX_STRING, pos, flags) setInfo refTpe
+
+      // used twice: once in the same owner as the lazy val, another time inside the compute method
+      val localLazyName = lazyName append LAZY_LOCAL_SUFFIX_STRING
+
+      // The lazy holder val need not be mutable, as we write to its field.
+      // In fact, it MUST not be mutable to avoid capturing it as an ObjectRef in lambdalift
+      // Must be marked LAZY to allow forward references, as in `def test2 { println(s.length) ; lazy val s = "abc" }
+      val holderSym = owner.newValue(localLazyName, pos, LAZY | ARTIFACT) setInfo refTpe
 
       val initializedGetter = refTpe.member(nme.initialized)
       val isUnit            = refClass == LazyUnitClass
       val valueGetter       = if (isUnit) NoSymbol else refTpe.member(nme.value)
-      val getValue          = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
+      def getValue          = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
 
-      def mkChecked(res: Tree) = If(Ident(holderSym) DOT initializedGetter, getValue, res)
+      def mkChecked(alreadyComputed: Tree, compute: Tree) = If(Ident(holderSym) DOT initializedGetter, alreadyComputed, compute)
 
       val computerSym =
         owner.newMethod(lazyName append nme.LAZY_SLOW_SUFFIX, pos, ARTIFACT | PRIVATE) setInfo MethodType(Nil, lazyValType)
 
       val rhsAtComputer = rhs.changeOwner(lazyVal -> computerSym)
       val computer = mkAccessor(computerSym) {
         val setInitialized = Apply(Select(Ident(holderSym), initializedGetter.setterIn(refClass)), TRUE :: Nil)
-        val setValue =
-          if (isUnit) rhsAtComputer
-          else Apply(Select(Ident(holderSym), valueGetter.setterIn(refClass)), rhsAtComputer :: Nil)
 
-        gen.mkSynchronized(Ident(holderSym))(mkChecked(Block(List(setValue, setInitialized), getValue)))
+        // we write to a local var outside of the synchronized block to avoid boxing/unboxing (synchronized method is polymorphic)
+        val resVarSym = computerSym.newVariable(localLazyName, pos, ARTIFACT) setInfo lazyValType
+
+        if (isUnit)
+          gen.mkSynchronized(Ident(holderSym))(mkChecked(alreadyComputed = UNIT, compute = Block(List(rhsAtComputer, setInitialized), UNIT)))
+        else {
+          val alreadyComputed = Assign(Ident(resVarSym), getValue)
+          val storeComputed   = Apply(Select(Ident(holderSym), valueGetter.setterIn(refClass)), Ident(resVarSym) :: Nil)
+          val compute = Block(List(Assign(Ident(resVarSym), rhsAtComputer), storeComputed, setInitialized), Ident(resVarSym))
```

Comment:
Looks like it's because the expected type for the `if` inside the `synchronized` is `Object` (the erased type of its type param `T`), and thus the branches are adapted to `Object` instead of `Unit`. Maybe we should give `synchronized` special treatment like `isInstanceOf`